Delete offline restclient from clusterKubeClientMap

FengyunPan opened the pull request:
When federation controller manager checks cluster status, it will
delete the offline cluster from clusterSet, but do not delete the
corresponding restclient from the map clusterKubeClientMap for
the offline cluster. This patch can fix it.

<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://github.com/kubernetes/kubernetes/blob/master/CONTRIBUTING.md and developer guide https://github.com/kubernetes/kubernetes/blob/master/docs/devel/development.md
2. If you want *faster* PR reviews, read how: https://github.com/kubernetes/kubernetes/blob/master/docs/devel/faster_reviews.md
3. Follow the instructions for writing a release note: https://github.com/kubernetes/kubernetes/blob/master/docs/devel/pull-requests.md#release-notes
-->

**What this PR does / why we need it**:

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Special notes for your reviewer**:

**Release note**:
<!--  Steps to write your release note:
1. Use the release-note-* labels to set the release note state (if you have access) 
2. Enter your extended release note in the below block; leaving it blank means using the PR title as the release note. If no release note is required, just write `NONE`. 
-->
```release-note
```

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/42567)
<!-- Reviewable:end -->

FengyunPan commented:
I signed it!

nikhiljindal commented:
This code creates and deleted clientsets from the map only when a new cluster is registered or unregistered. It does not change kubeClientMap when a cluster goes offline or comes back up.

What is the motivation for changing that? Is there a bug you are seeing that you want to fix? Or looking to optimize the code?
If we were to accept this change, we will need a change to create the clientset in line 145 as well.

FengyunPan commented:
@nikhiljindal Thank for your review.
when ClusterController observed a new cluster, it adds cluster.name into knownClusterSet in line 146 and adds clusterClient into clusterKubeClientMap in line 131, But when ClusterController observed a Cluster deletion, it just delete cluster.name from knownClusterSet in line 161.

FengyunPan commented:
I will create new patch to fix it and optimize the code.

FengyunPan commented:
@huangyuqi Hello huangyuqi, I'd like you to do a code review, thank you.

nikhiljindal commented on federation/pkg/federation-controller/cluster/clustercontroller.go:
```diff
@@ -115,18 +115,28 @@ func (cc *ClusterController) Run() {
 }
 
-func (cc *ClusterController) GetClusterStatus(cluster *federationv1beta1.Cluster) (*federationv1beta1.ClusterStatus, error) {
+// GetClusterClient gets cluster client.
```
Remove comment. Not adding any value

FengyunPan replied:
Done

nikhiljindal commented on federation/pkg/federation-controller/cluster/clustercontroller.go:
```diff
@@ -115,18 +115,28 @@ func (cc *ClusterController) Run() {
+	return client, nil
+}
+
+// GetClusterStatus gets cluster status.
```
same. comment is not adding any value

FengyunPan replied:
Done

nikhiljindal commented on federation/pkg/federation-controller/cluster/clustercontroller.go:
```diff
@@ -141,7 +151,13 @@ func (cc *ClusterController) UpdateClusterStatus() error {
 	for _, cluster := range clusters.Items {
 		if !cc.knownClusterSet.Has(cluster.Name) {
 			glog.V(1).Infof("ClusterController observed a new cluster: %#v", cluster)
+			// add cluster name into knownClusterSet and add cluster client into clusterKubeClientMap
```
How about calling addToClusterSet here?
You can also move the glog.V(1) new cluster observed to that function

FengyunPan replied:
Good idea.

FengyunPan replied:
Good idea. Done

nikhiljindal commented on federation/pkg/federation-controller/cluster/clustercontroller.go:
```diff
@@ -154,7 +170,9 @@ func (cc *ClusterController) UpdateClusterStatus() error {
 		deleted := cc.knownClusterSet.Difference(observedSet)
 		for clusterName := range deleted {
 			glog.V(1).Infof("ClusterController observed a Cluster deletion: %v", clusterName)
+			// delete the cluster name from knownClusterSet and delete the cluster client from clusterKubeClientMap
```
same. how about calling delFromClusterSet here? You can also move glog.V(1) observed a cluster deletion to that function

FengyunPan replied:
Good idea. Done

nikhiljindal commented on federation/pkg/federation-controller/cluster/clustercontroller.go:
```diff
@@ -85,14 +85,24 @@ func NewclusterController(federationClient federationclientset.Interface, cluste
 // delete the corresponding restclient from the map clusterKubeClientMap
 func (cc *ClusterController) delFromClusterSet(obj interface{}) {
 	cluster := obj.(*federationv1beta1.Cluster)
+	glog.V(1).Infof("ClusterController observed a cluster deletion: %v", cluster.Name)
```
call delFromClusterSetByName here rather than duplicating code.

FengyunPan replied:
Done.

FengyunPan replied:
v1beta1.Cluster

FengyunPan replied:
Done.

nikhiljindal commented on federation/pkg/federation-controller/cluster/clustercontroller.go:
```diff
@@ -115,18 +125,26 @@ func (cc *ClusterController) Run() {
 
-func (cc *ClusterController) GetClusterStatus(cluster *federationv1beta1.Cluster) (*federationv1beta1.ClusterStatus, error) {
+func (cc *ClusterController) GetClusterClient(cluster *federationv1beta1.Cluster) (*ClusterClient, error) {
 	// just get the status of cluster, by requesting the restapi "/healthz"
```
remove this comment from here. Not relevant for this new function

FengyunPan replied:
Done.

FengyunPan replied:
Done.

nikhiljindal commented:
Thanks for the fixes @FengyunPan 
Should be the last round of comments.

@k8s-bot ok to test

nikhiljindal commented:
/release-note-none

Commit:
Delete cluster client for a Cluster deletion

When federation controller manager checks cluster status, it will
delete the offline cluster from clusterSet, but do not delete the
corresponding restclient from the map clusterKubeClientMap for
the offline cluster. This patch can fix it.

FengyunPan commented:
@k8s-bot bazel test this
@k8s-bot unit test this
@k8s-bot kubemark e2e test this

nikhiljindal commented:
Thanks for the fixes.
/lgtm
/approve

@k8s-bot kops aws e2e test this
